include/ginkgo/core/base/range_accessor_helper.hpp

```diff
@@ -0,0 +1,586 @@
+ * This helper runs from first to last dimension in order to compute the index.
+ * The index is computed like this:
+ * indices: x1, x2, x3, ...
+ * compute(stride, x1, x2, x3) -> x1 * stride[0] + x2 * stride[1] + x3
+ */
+template <typename ValueType, size_type total_dim, size_type current_iter = 1>
+struct row_major_helper_s {
+    static_assert(total_dim >= 1, "Dimensionality must be >= 1");
+    static_assert(current_iter < total_dim, "Iteration must be < total_dim!");
+
+    static constexpr size_type dim_idx{current_iter - 1};
+
+    template <typename FirstType, typename... Indices>
+    static constexpr GKO_ATTRIBUTES GKO_INLINE ValueType
+    compute(const dim<total_dim> &size,
+            const std::array<ValueType, (total_dim > 1 ? total_dim - 1 : 0)>
+                &stride,
+            FirstType &&first, Indices &&... idxs)
+    {
+        // The ASSERT size check must NOT be indexed with `dim_idx` directy,
+        // otherwise, it leads to a linker error. The reason is likely that
+        // `dim<N>::operator[](const size_type &)` uses a reference. Since
+        // `dim_idx` is constexpr (and not defined in a namespace scope), it
+        // can't be odr-used.
+        return GKO_ASSERT(first < size[static_cast<size_type>(dim_idx)]),
+               first * stride[dim_idx] +
+                   row_major_helper_s<ValueType, total_dim, current_iter + 1>::
+                       compute(size, stride, std::forward<Indices>(idxs)...);
+    }
+};
+
+template <typename ValueType, size_type total_dim>
+struct row_major_helper_s<ValueType, total_dim, total_dim> {
+    template <typename FirstType>
+    static constexpr GKO_ATTRIBUTES GKO_INLINE ValueType
+    compute(const dim<total_dim> &size,
+            const std::array<ValueType, (total_dim > 1 ? total_dim - 1 : 0)>,
+            FirstType &&first)
+    {
+        return GKO_ASSERT(first < size[total_dim - 1]), first;
+    }
+};
+
+
+template <typename ValueType, size_type iter, size_type N>
+constexpr GKO_ATTRIBUTES GKO_INLINE std::enable_if_t<iter == N, ValueType>
+mult_array(const dim<N> &size)
+{
+    return 1;
+}
+
+template <typename ValueType, size_type iter, size_type N>
+constexpr GKO_ATTRIBUTES GKO_INLINE std::enable_if_t<(iter < N), ValueType>
+mult_array(const dim<N> &size)
+{
+    return size[iter] * mult_array<ValueType, iter + 1>(size);
+}
+
+
+template <typename ValueType, size_type iter = 1, size_type N, typename... Args>
+constexpr GKO_ATTRIBUTES GKO_INLINE
+    std::enable_if_t<N == 0 || (iter == N && iter == sizeof...(Args) + 1),
+                     std::array<ValueType, N == 0 ? 0 : N - 1>>
+    extract_factorization(const dim<N> &size, Args... args)
+{
+    return {{args...}};
+}
+
+template <typename ValueType, size_type iter = 1, size_type N, typename... Args>
+constexpr GKO_ATTRIBUTES GKO_INLINE std::enable_if_t<
+    (iter < N) && (iter == sizeof...(Args) + 1), std::array<ValueType, N - 1>>
+extract_factorization(const dim<N> &size, Args... args)
```

Comment:
I'm again not sure about the naming here. What this does is call `mult_array` for every dimension successively, carry all those results around, until it's stored in an array with the previous call. I would name this `compute_default_strides` or something like that.

Reply:
Excellent point, I renamed this to `compute_default_stride_array`.
I actually meant the opposite of `factorization` (`extract_recursive_product` might have been a bit more accurate), but I agree that even that is not stating the purpose of the function.